crypto/fipsmodule/ec/ec_nistp.c

```diff
@@ -498,3 +498,141 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
+//
+// The group (0) digits correspond precisely to the multiples of G that are
+// held in the 27 precomputed sub-tables, so we may simply read the appropriate
+// points from the sub-tables and sum them all up (negating if needed, i.e., if
+// a digit s_i is negative, we read the point corresponding to the abs(s_i) and
+// negate it before adding it to the sum).
+// The remaining three groups (1), (2), and (3), correspond to the multiples
+// of G from the sub-tables multiplied additionally by 2^5, 2^10, and 2^15,
+// respectively. Therefore, for these groups we may read the appropriate points
+// from the table, double them 5, 10, or 15 times, respectively, and add them
+// to the final result.
+//
+// To minimize the number of required doubling operations we process the digits
+// of the scalar from left to right. In other words, the algorithm is:
+//   1. For group (i) in this order (3, 2, 1, 0):
+//   2.   Double the accumulator 5 times except in the first iteration.
+//   3.   Read the points corresponding to the group (i) digits from the tables
+//        and add them to an accumulator.
+//   4. If the scalar is even subtract G from the accumulator.
+//
+// Note: this function is designed to be constant-time.
+void ec_nistp_scalar_mul_base(const ec_nistp_meth *ctx,
+                              ec_nistp_felem_limb *x_out,
+                              ec_nistp_felem_limb *y_out,
+                              ec_nistp_felem_limb *z_out,
+                              const EC_SCALAR *scalar) {
+  // Regular-wNAF encoding of the scalar.
+  int16_t rwnaf[SCALAR_MUL_MAX_NUM_WINDOWS];
+  scalar_rwnaf(rwnaf, SCALAR_MUL_WINDOW_SIZE, scalar, ctx->felem_num_bits);
+  size_t num_windows = DIV_AND_CEIL(ctx->felem_num_bits, SCALAR_MUL_WINDOW_SIZE);
+
+  // We need two point accumulators, so we define them of maximum size
+  // to avoid allocation, and just take pointers to individual coordinates.
+  // (This cruft will disapear when we refactor point_add/dbl to work with
+  // whole points instead of individual coordinates).
+  ec_nistp_felem_limb res[3 * FELEM_MAX_NUM_OF_LIMBS] = {0};
+  ec_nistp_felem_limb tmp[3 * FELEM_MAX_NUM_OF_LIMBS] = {0};
+  ec_nistp_felem_limb *x_res = &res[0];
+  ec_nistp_felem_limb *y_res = &res[ctx->felem_num_limbs];
+  ec_nistp_felem_limb *z_res = &res[ctx->felem_num_limbs * 2];
+  ec_nistp_felem_limb *x_tmp = &tmp[0];
+  ec_nistp_felem_limb *y_tmp = &tmp[ctx->felem_num_limbs];
+  ec_nistp_felem_limb *z_tmp = &tmp[ctx->felem_num_limbs * 2];
+
+  // Process the 4 groups of digits starting from group (3) down to group (0).
+  for (int i = 3; i >= 0; i--) {
+    // Double |res| 5 times in each iteration, except in the first one.
+    for (size_t j = 0; i != 3 && j < SCALAR_MUL_WINDOW_SIZE; j++) {
+      ctx->point_dbl(x_res, y_res, z_res, x_res, y_res, z_res);
+    }
+
+    // Process the digits in the current group from the most to the least
+    // significant one.
```

Comment:
Should we keep the original relevant comment? It does seem a bit important, but I'm not 100% sure.
```suggestion
    // Process the digits in the current group from the most to the least
    // significant one (this is a requirement to ensure that the case of point
    // doubling can't happen).
```

Reply:
https://github.com/aws/aws-lc/pull/2003#discussion_r1852681259